### ISSUE TITLE
BelongsToMany ordering with eager loading

### ISSUE THREAD

**Issue by ThomasGonet** [yajra/laravel-datatables#1094]
Hi all,

I'am facing a weird behaviour using ordering on belongstomany eager loading.

No problem for loading initial data, but order on column with relation doesn't work.

Controller part :
```php
public function anyData()
    {
        $devices = Device::with('model', 'users')->selectRaw('distinct devices.*');

        return Datatables::of($devices)
            ->addColumn('user', function (Device $device) {
                return $device->users
                    ->map(function($user) {
                    return $user->first_name.' '.$user->last_name ;
                })->implode('<br>');
            })
            ->make('true');
    }
```

Javascript part : 
```js
$(document).ready(function () {
    let rows_selected = [];
    let oTable = $('#devices').DataTable({
        "ajax": '/admin/devices/data',
        "paging": true,
        "lengthChange": true,
        "searching": true,
        "ordering": true,
        "info": true,
        "autoWidth": false,
        "serverSide": true,
        "processing": true,
        "searchable": true,
...
        "language": {
            "url": "/js/dist/french_datatables.json"
        },
        "columnDefs": [
...
            {
                'targets': 5,
                'searchable': false,
                'data': 'user',
                'name': 'users.id',
                'defaultContent': '',
            },
...
```

And the error given by browser 
```
QueryException in Connection.php line 647:
SQLSTATE[42S22]: Column not found: 1054 Unknown column 'users.users' in 'field list' (SQL: select distinct devices.*, `users`.`users` from `devices` left join `device_user` on `device_user`.`device_id` = `devices`.`id` left join `users` on `device_user`.`user_id` = `users`.`id` where `devices`.`deleted_at` is null order by `users`.`id` asc limit 10 offset 0)
```

It seems that a *`users`.`users`* column is automatically added but i'm not able to find why.

I used [this example](https://datatables.yajrabox.com/relation/belongs-to-many) to base my code.

EDIT :

Model and relations

Device: 

```php

Device extends Model {

	public $timestamps = false;
    protected $table = 'devices';

	use SoftDeletes;
	protected $dates = ['deleted_at'];
    protected $fillable = array('name', 'softDeletes', 'serial_number', 'active', 'frequency', 'model_device_id');
...
	public function model()
	{
        return $this->belongsTo('App\Models\ModelDevice', 'model_device_id', 'id');
	}

	public function users()
	{
        return $this->belongsToMany('App\Models\User')->withPivot("end_date", "state")->withTimestamps()->wherePivot('end_date', null);
	}
...
}
```

```php
....
User extends Authenticatable
{
...
	public $timestamps = true;
    protected $table = 'users';
    protected $fillable = array('first_name', 'last_name', 'phone', 'email', 'username', 'password', 'group_id', 'exploitation_id', 'shop_id', 'created_by', 'description');

...
}
```

System :
- Homestead vagrant
- PHP 7.1
- Laravel 5.4
- Laravel-Datatables 7.3

Thanks in advance for your help.

**Comment by 1147832740** [comment]
I have the same problem as you, it hasn't been solved yet.

**Comment by yajra** [comment]
I think this [commit](https://github.com/yajra/laravel-datatables/commit/984e033a8c8bd59075afd6e92748f486cc00c109) was causing this issue and needs some tweak.

```
Add eager loaded column to select.
Fix ORDER BY clause is not in SELECT list when MySQL is in strict mode.
```

**Comment by yajra** [comment]
Can you please paste your js columns definitions?

**Comment by ThomasGonet** [comment]
Thanks for your replies.


Here is my columns defs :
```javascript
 "columnDefs": [
            {
                'targets': 0,
                'searchable': false,
                'orderable': false,
                'width': '1%',
                'className': 'dt-body-center thead_checkbox',
                'render': function (data, type, full, meta) {
                    return '<input type="checkbox">';
                }
            },
            {
                'targets': 1,
                'searchable': true,
                'orderable': true,
                'data': 'serial_number',
                'name': 'devices.serial_number',
            },
            {
                'targets': 2,
                'searchable': true,
                'data': 'name',
                'name': 'devices.name',
                'defaultContent': '',
            },
            {
                'targets': 3,
                'searchable': true,
                'data': 'model.name',
                'name': 'devices.model.name',
                'defaultContent': '',
            },
            {
                'targets': 4,
                'searchable': false,
                'data': 'frequency',
                'name': 'devices.frequency',
                'defaultContent': '',
            },
            {
                'targets': 5,
                'searchable': false,
                'orderable': true,
                'data': 'user',
                'name': 'users.id',
                'defaultContent': '',
            },
            {
                'targets': 6,
                'searchable': false,
                'orderable': false,
                'className': '',
                'width': '12%',
                'render': function (data, type, full, meta) {
                    let button = '<a href="/devices/' + full.serial_number + '"><button class="btn btn-default btn-block" type="button">Détails</button></a>';
                    return button;
                }
            },
            {
                'targets': 7,
                'searchable': false,
                'orderable': false,
                'className': '',
                'width': '12%',
                'render': function (data, type, full, meta) {
                    let button = '<button class="btn btn-danger btn-block" type="button" data-toggle="modal" data-target="#delete-modal" data-id="' + full.id + '" data-serial-number="' + full.serial_number + '">Supprimer</button>';
                    return button;
                }
            },
        ],
```

Trying in command line for mysql seems it need *users.id* column.

I will try to look on the QueryBuilderEngine but I don't think I can fix this myself...

**Comment by stefanocurnis** [comment]
Hey guys.

Same problem here.
Here's my code:

### Models and Relationships

```
class Product extends Model
{

    ...

    public function categories() {
        return $this->belongsToMany('App\Models\ProductCategory','product_categories_products');
    }
}
```


### Javascript column definition

```
"columns": [
     ...
     {
     "name": "categories.name",
     "data": "categories",
     "title": "Categorie",
     "orderable": true,
     "searchable": true,
     "attributes": []
     },
     ....
]
```

### Controller code

```
        if ($request->ajax()) {

            $query=Product::with('categories')->select('products.*');



            return Datatables::of($query)
                ...

               ->addColumn('categories', function (Product $product) {
                   return $product->categories->map(function($category) {
                       return str_limit($category->name, 30, '...');
                   })->implode('<br>');
               })
                ...
                ->rawColumns(['price','categories','enabled', 'action'])
                ->make(true);

        }

```


## How to reproduce

1. Load the page normally. Datatable works as expected. No errors.
2. Search for something. It works, even by `categories column`
3. Order by `categories column`. **Result: AJAX Error**
4. Search no longer works.

**AJAX / Eloquent / SQL Error.**
`Column not found: 1054 Unknown column 'product_categories.categories' in 'field list' (SQL: select 'products'.*, 'product_categories'.'categories' from 'products' left join 'product_categories_products'`

Thanks guys! I'll try to fix and let you know...

**Comment by yajra** [comment]
Should be fixed on the next release v7.5.1++ Thanks!
